### PR TITLE
Add JetService, deprecate Jet, JetInstance

### PR DESCRIPTION
**Pull request by ufukyilmaz**
This PR deprecates `JetInstance` and `Jet` interfaces
and introduces a new public interface called `JetService`
replacing the usages of deprecated `JetInstance`. It aims
to approach Jet as an extension service to `HazelcastInstance`
instead of being an instance on its own which encapsulates
`HazelcastInstance`. `JetService` does not have `#getMap()`
like data structure accessors (except jet observables) and does
not have cluster management methods such as `#shutdown()`,
etc.. Thus, from now on, cluster operations and data structure
accesses (except Jet observables) can only be done through
 `HazelcastInstance`.

This PR also adds back previously removed methods in `Jet`
class such as `Jet#newJetClient` and deprecates them.

Related EE PR: https://github.com/hazelcast/hazelcast-enterprise/pull/4068

Breaking changes: 
- Added `HazelcastInstance hazelcastInstance()` which has no
default impl. into `ProcessMetaSupplier.Context`
- Move the default implementations of methods such as
`getJobStateSnapshots(String)` from `JetInstance` to
 `AbstractJetInstance`. These default implementations already
contained `AbstractJetInstance` accesses by using downcastings
to `this` JetInstance into `AbstractJetInstance` etc., which previously
made the `JetInstance` impossible to be extended without extending
`AbstractJetInstance`. So, we don't need to care about it.

Deprecated staff:
- `Jet` class and all of static factory methods in it are deprecated
`JetInstance` class deprecated and replaced by `JetService`. The
 methods of `JetInstance` which are not migrated to `JetService`
 are also deprecated.
- `ProcessMetaSupplier#Context.jetInstance()` is deprecated.
- `JetCacheManager` is deprecated.

Missing things that I don't want to include the PR's scope are:
- Add configuration for creating `JetInstance` spring bean. In this
PR, I replaced the `JetInstance` bean configuration to create 
`JetService` bean instead of `JetInstance`. But after that, I didn't
add any other bean configurations for creating the `JetInstance` bean.
- Remove the `JetInstance` usages from Jet tests
- Update `JetCommandLine`

Closes:  #18787

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Add `Add to Release Notes` label if changes should be mentioned in release notes or `Not Release Notes content` if changes are not relevant for release notes
- [x] Request reviewers if possible
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [ ] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

**Comment by ufukyilmaz** [comment]
This PR is ready for initial review. I'd like API changes to be reviewed.

File: hazelcast/src/main/java/com/hazelcast/jet/Jet.java [hunk 1]
```diff
@@ -27,6 +35,7 @@
  *
  * @since 3.0
  */
+@Deprecated
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/Jet.java:38]: Each deprecated class/method should have a javadoc pointing the user to the replacement. If the new method fully replaces the deprecated one, then we should also remove the entire javadoc from the old method.

File: hazelcast/src/main/java/com/hazelcast/jet/Jet.java [hunk 2]
```diff
@@ -100,9 +109,91 @@ private Jet() {
      * </pre>
      *
      * @since 4.0
+     * @deprecated
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/Jet.java:112]: Remove the javadoc?

File: hazelcast/src/main/java/com/hazelcast/jet/Jet.java [hunk 3]
```diff
@@ -100,9 +109,91 @@ private Jet() {
...
+    @Deprecated
     @Nonnull
     public static JetInstance bootstrappedInstance() {
         return JetBootstrap.getInstance();
     }
+
+    /**
+     * Creates a member of the Jet cluster with the configuration loaded from
+     * default location.
+     * @deprecated see {@linkplain Hazelcast#newHazelcastInstance()}
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/Jet.java:123]:
```suggestion
     * @deprecated Use {@link Hazelcast#newHazelcastInstance().getJet()} instead.
```

File: hazelcast/src/main/java/com/hazelcast/jet/JetService.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.List;
+
+import static com.hazelcast.jet.impl.JobRepository.exportedSnapshotMapName;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Represents either an instance of a
+ * instance that connects to a remote cluster.
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/JetService.java:41]: Unfinished comment?

Now it doesn't represent an instance, but a service.

File: hazelcast/src/main/java/com/hazelcast/jet/JetService.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+ * @since 5.0
+ */
+public interface JetService {
+    /**
+     * Returns the underlying Hazelcast IMDG instance used by Jet. It will
+     * be either a server node or a client, depending on the type of this
+     * {@code JetInstance}.
+     */
+    @Nonnull
+    HazelcastInstance getHazelcastInstance();
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/JetService.java:51]: Do we need this method here? We don't have backlinks from other services, e.g. from `SqlService` or from `CPSubsystem`.

File: hazelcast/src/main/java/com/hazelcast/jet/JobStateSnapshot.java [hunk 6]
```diff
@@ -32,16 +33,23 @@
...
     private final SnapshotValidationRecord snapshotValidationRecord;
 
-    JobStateSnapshot(@Nonnull JetInstance instance, @Nonnull String name, @Nonnull SnapshotValidationRecord record) {
+    JobStateSnapshot(@Nonnull HazelcastInstance instance, @Nonnull String name, @Nonnull SnapshotValidationRecord record) {
         this.instance = instance;
         this.name = name;
         this.snapshotValidationRecord = record;
     }
 
+    @Deprecated
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/JobStateSnapshot.java:46]: This is private method, we don't need to keep the deprecated one.

File: hazelcast/src/main/java/com/hazelcast/jet/impl/util/Util.java [hunk 7]
```diff
@@ -137,7 +136,7 @@ public static boolean tryIncrement(AtomicInteger value, int increment, int limit
     }
 
     public static JetInstance getJetInstance(NodeEngine nodeEngine) {
-        return nodeEngine.<JetService>getService(JetService.SERVICE_NAME).getJetInstance();
+        return nodeEngine.<JetServiceBackend>getService(JetServiceBackend.SERVICE_NAME).getJetInstance();
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/util/Util.java:139]: I guess we could make this method return the new `JetService` to stop using the deprecated APIs internally.

File: hazelcast/src/main/java/com/hazelcast/jet/server/JetCommandLine.java [hunk 8]
```diff
@@ -176,7 +176,7 @@ public JetCommandLine(Function<ClientConfig, JetInstance> jetClientFn, PrintStre
 
     public static void main(String[] args) {
         runCommandLine(
-                config -> HazelcastClient.newHazelcastClient(config).getJetInstance(),
+                config -> (JetInstance) HazelcastClient.newHazelcastClient(config).getJet(),
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/server/JetCommandLine.java:179]: The `runCommandLine` could be changed to accept `JetService`

File: hazelcast/src/main/java/com/hazelcast/jet/server/JetCommandLine.java [hunk 9]
```diff
@@ -114,8 +114,8 @@
 
 @SuppressWarnings({"unused", "MismatchedQueryAndUpdateOfCollection"})
 @Command(
-        name = "jet",
-        description = "Utility to perform operations on a Hazelcast Jet cluster.%n" +
+        name = "platform",
+        description = "Utility to perform operations on a Hazelcast Platform cluster.%n" +
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/server/JetCommandLine.java:118]: I've heard that the new merged thing will be called just "Hazelcast". Is the `name` used anywhere on command line? Can't we remove it completely?
```suggestion
        name = "hazelcast",
        description = "Utility to perform operations on a Hazelcast cluster.%n" +
```

**Review by viliam-durina — COMMENTED** [review]
I reviewed only the public API, not any other changes.

**Comment by viliam-durina** [comment]
Missing pieces:
- `ProcessorMetaSupplier.jetInstance()` - I think we should add `hazelcastInstance()`, not `jetService()`
- `TestSupport.jetInstance()`
- many examples in javadocs still use `JetInstance` - search for this keyword in comments

File: hazelcast/src/main/java/com/hazelcast/jet/server/JetCommandLine.java (hunk 8 above)
Comment (ufukyilmaz) [reply hazelcast/src/main/java/com/hazelcast/jet/server/JetCommandLine.java:179]: I am thinking of converting this class to `HazelcastCommandLine` class. I asked it in slack. If it is decided, I will do it in a separate PR.

File: hazelcast/src/main/java/com/hazelcast/jet/server/JetCommandLine.java (hunk 9 above)
Comment (ufukyilmaz) [reply hazelcast/src/main/java/com/hazelcast/jet/server/JetCommandLine.java]: It is fixed after merging master into this branch.

File: hazelcast/src/main/java/com/hazelcast/jet/JobStateSnapshot.java (hunk 6 above)
Comment (ufukyilmaz) [reply hazelcast/src/main/java/com/hazelcast/jet/JobStateSnapshot.java]: I removed it.

File: hazelcast/src/main/java/com/hazelcast/jet/JetService.java (hunk 5 above)
Comment (ufukyilmaz) [reply hazelcast/src/main/java/com/hazelcast/jet/JetService.java]: I removed this backlink.

**Comment by ufukyilmaz** [comment]
Missing things:
- Spring JetInstance support (I replaced JetInstance bean with JetService with breaking the compatibility.) I don't know how to add old JetInstance again.
- Decide what to do with Jet observables
- Update Jet test to not use JetInstance
- Consider removing `InstanceConfig`, https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/jet/config/InstanceConfig.java.
- Update JetCommandLine (which will be in the scope another PR.)

**Comment by ufukyilmaz** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/jet/impl/JobRepository.java [hunk 10]
```diff
@@ -537,15 +535,15 @@ private boolean isResourceMapExpired(long creationTime) {
 
     private Map<Long, JobRecord> jobRecordsMap() {
         if (jobRecords.remembered() != null ||
-                ((AbstractJetInstance) jetInstance).existsDistributedObject(SERVICE_NAME, JOB_RECORDS_MAP_NAME)) {
+                ((AbstractJetInstance) instance.getJet()).existsDistributedObject(SERVICE_NAME, JOB_RECORDS_MAP_NAME)) {
```
[the remark this answers is not in the document]
Comment (ufukyilmaz) [reply hazelcast/src/main/java/com/hazelcast/jet/impl/JobRepository.java:538]: Moving these `existsDistributedObject()` method requires also changing the client protocol as well if we want consistent naming. See used codec: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/jet/impl/JetClientInstanceImpl.java#L115-L121. I don't want to include this change into this PR. wdyt?

File: hazelcast/src/main/java/com/hazelcast/jet/JetCacheManager.java [hunk 11]
```diff
@@ -17,49 +17,19 @@
...
- * use of reflection and/or dynamic proxies (e.g. Mockito, Spring etc)
- * to operate on {@link JetInstance} when JCache is not on the classpath.
- *
- * @since 3.0
+ * @since Jet 3.0
+ * @deprecated since 5.0
+ * Use {@link com.hazelcast.core.ICacheManager} instead.
  */
+@Deprecated
 public interface JetCacheManager {
```
[the remark this answers is not in the document]
Comment (ufukyilmaz) [reply hazelcast/src/main/java/com/hazelcast/jet/JetCacheManager.java:27]: I think it'd better drop it along with `JetInstance` removal to not to break.

File: hazelcast/src/main/java/com/hazelcast/jet/JetInstance.java [hunk 12]
```diff
@@ -18,461 +18,174 @@
...
+     * @deprecated since 5.0
+     * We left it here since it has a default implementation depending
+     * on {@link JetInstance#getHazelcastInstance()}.
+     * Prefer to use {@link JetService#getJobStateSnapshot(String)}
+     * instead.
      */
     @Nullable
+    @Override
+    @Deprecated
     default JobStateSnapshot getJobStateSnapshot(@Nonnull String name) {
```
[the remark this answers is not in the document]
Comment (ufukyilmaz) [reply hazelcast/src/main/java/com/hazelcast/jet/JetInstance.java]: I moved these implementations down to `AbstractJetInstance`.

File: hazelcast/src/main/java/com/hazelcast/jet/Jet.java [hunk 13]
```diff
@@ -37,12 +44,88 @@ private Jet() {
...
+
+    /**
+     * @deprecated since 5.0
+     * Use {@link Hazelcast#newHazelcastInstance(Config)} and then get
+     * {@link JetService} from the created {@link HazelcastInstance} by
+     * using {@link HazelcastInstance#getJet()}}.
+     */
+    @Deprecated
+    @Nonnull
+    public static JetInstance newJetInstance(@Nonnull JetConfig config) {
```
[the remark this answers is not in the document]
Comment (ufukyilmaz) [reply hazelcast/src/main/java/com/hazelcast/jet/Jet.java:76]: Fixed it.

**Comment by viliam-durina** [comment]
There are still plenty of `@since` javadocs in Jet packages that don't say `@Since Jet x.y`
image

File: hazelcast/src/main/java/com/hazelcast/jet/impl/util/Util.java [hunk 14]
```diff
@@ -647,7 +646,4 @@ public static String formatJobDuration(long durationMs) {
         };
     }
 
-    public static NodeEngineImpl getNodeEngine(JetInstance instance) {
-        return ((HazelcastInstanceImpl) instance.getHazelcastInstance()).node.nodeEngine;
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/util/Util.java:651]: You could have kept this method, the code in it quite bad-looking and it was used a couple of times.

**Review by viliam-durina — APPROVED** [review]

**Review by mdumandag — APPROVED** [review]
OK for the client-side changes

File: hazelcast/src/main/java/com/hazelcast/jet/impl/JetServiceBackend.java [hunk 15]
```diff
@@ -58,9 +59,9 @@
 import static com.hazelcast.jet.impl.util.Util.memoizeConcurrent;
 import static java.util.concurrent.TimeUnit.SECONDS;
 
-public class JetService implements ManagedService, MembershipAwareService, LiveOperationsTracker {
+public class JetServiceBackend implements ManagedService, MembershipAwareService, LiveOperationsTracker {
 
-    public static final String SERVICE_NAME = "hz:impl:jetService";
+    public static final String SERVICE_NAME = "hz:impl:jetServiceBackend";
```
Comment (gurbuzali) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/JetServiceBackend.java:64]: why change the internal name ? I don't think we should add `Backend` here

**Review by gurbuzali — APPROVED** [review]

**Comment by ufukyilmaz** [comment]
run-lab-run

**Review by frant-hartm — APPROVED** [review]

**Comment by ufukyilmaz** [comment]
run-lab-run

**Pull request merged**